kamu-core/src/infra/repos/dataset_repository_s3.rs

```diff
@@ -0,0 +1,473 @@
+        let handle = self.resolve_dataset_ref(dataset_ref).await?;
+        let dataset = self.get_dataset_impl(&handle.name)?;
+        Ok(Arc::new(dataset))
+    }
+
+    async fn create_dataset(
+        &self,
+        dataset_name: &DatasetName,
+    ) -> Result<Box<dyn DatasetBuilder>, BeginCreateDatasetError> {
+        let staging_name = get_staging_name();
+        let dataset_url = self.get_s3_bucket_path(staging_name.as_str());
+        let dataset_result = DatasetFactoryImpl::get_s3(dataset_url);
+        match dataset_result {
+            Ok(dataset) => Ok(Box::new(DatasetS3BuilderImpl::new(
+                self.clone(),
+                Arc::new(dataset),
+                staging_name,
+                dataset_name.clone(),
+            ))),
+            Err(e) => Err(BeginCreateDatasetError::Internal(e)),
+        }
+    }
+
+    async fn rename_dataset(
+        &self,
+        dataset_ref: &DatasetRefLocal,
+        new_name: &DatasetName,
+    ) -> Result<(), RenameDatasetError> {
+        let old_name = self.resolve_dataset_ref(dataset_ref).await?.name;
+        match self
+            .move_bucket_items_on_dataset_rename(old_name.as_str(), new_name.as_str())
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(MoveBucketItemsOnRenameError::NameCollision(e)) => {
+                Err(RenameDatasetError::NameCollision(e))
+            }
+            Err(MoveBucketItemsOnRenameError::Internal(e)) => Err(RenameDatasetError::Internal(e)),
+        }
+    }
+
+    async fn delete_dataset(
+        &self,
+        dataset_ref: &DatasetRefLocal,
+    ) -> Result<(), DeleteDatasetError> {
+        let dataset_handle = match self.resolve_dataset_ref(dataset_ref).await {
+            Ok(dataset_handle) => dataset_handle,
+            Err(GetDatasetError::NotFound(e)) => return Err(DeleteDatasetError::NotFound(e)),
+            Err(GetDatasetError::Internal(e)) => return Err(DeleteDatasetError::Internal(e)),
+        };
+
+        let children: Vec<_> = get_downstream_dependencies_impl(self, dataset_ref)
+            .try_collect()
+            .await?;
+
+        if !children.is_empty() {
+            return Err(DanglingReferenceError {
+                dataset_handle,
+                children,
+            }
+            .into());
+        }
+
+        match self.eliminate_bucket_items(&dataset_handle.name).await {
+            Ok(_) => Ok(()),
+            Err(e) => Err(DeleteDatasetError::Internal(e)),
+        }
+    }
+
+    fn get_downstream_dependencies<'s>(
+        &'s self,
+        dataset_ref: &'s DatasetRefLocal,
+    ) -> DatasetHandleStream<'s> {
+        Box::pin(get_downstream_dependencies_impl(self, dataset_ref))
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+// DatasetS3BuilderImpl
+/////////////////////////////////////////////////////////////////////////////////////////
+
+struct DatasetS3BuilderImpl {
```

Comment:
mb `DatasetBuilderS3`?

Reply:
Renamed as suggested. Similarly, renamed the builder in the local FS version to `DatasetBuilderLocalFs`